src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeWriter.cs

```diff
@@ -0,0 +1,316 @@
+            sizeHint = Math.Max(_minimumBufferSize, sizeHint);
+            // After that adjust it to fit into pools max buffer size
+            var adjustedToMaximumSize = Math.Min(maxBufferSize, sizeHint);
+            return adjustedToMaximumSize;
+        }
+
+        private BufferSegment CreateSegmentUnsynchronized()
+        {
+            if (_bufferSegmentPool.TryPop(out BufferSegment segment))
+            {
+                return segment;
+            }
+
+            return new BufferSegment();
+        }
+
+        private void ReturnSegmentUnsynchronized(BufferSegment segment)
+        {
+            if (_bufferSegmentPool.Count < MaxSegmentPoolSize)
+            {
+                _bufferSegmentPool.Push(segment);
+            }
+        }
+
+        /// <inheritdoc />
+        public override void CancelPendingFlush()
+        {
+            Cancel();
+        }
+
+        /// <inheritdoc />
+        public override void Complete(Exception exception = null)
+        {
+            if (_isCompleted)
+            {
+                return;
+            }
+
+            _isCompleted = true;
+
+            _internalTokenSource?.Dispose();
+
+            BufferSegment segment = _head;
+            while (segment != null)
+            {
+                BufferSegment returnSegment = segment;
+                segment = segment.NextSegment;
+                returnSegment.ResetMemory();
+            }
+
+            _head = null;
+            _tail = null;
+
+            // REVIEW: Do we need a leaveOpen to avoid this?
+            InnerStream.Dispose();
+        }
+
+        /// <inheritdoc />
+        public override void OnReaderCompleted(Action<Exception, object> callback, object state)
+        {
+            throw new NotSupportedException("OnReaderCompleted is not supported.");
+        }
+
+        /// <inheritdoc />
+        public override ValueTask<FlushResult> FlushAsync(CancellationToken cancellationToken = default)
+        {
+            if (_bytesBuffered == 0)
+            {
+                return new ValueTask<FlushResult>(new FlushResult(isCanceled: false, isCompleted: false));
+            }
+
+            return FlushAsyncInternal(cancellationToken);
+        }
+
+        private void Cancel()
+        {
+            InternalTokenSource.Cancel();
+        }
+
+        private async ValueTask<FlushResult> FlushAsyncInternal(CancellationToken cancellationToken = default)
+        {
+            // Write all completed segments and whatever remains in the current segment
+            // and flush the result.
+            var reg = new CancellationTokenRegistration();
+            if (cancellationToken.CanBeCanceled)
+            {
+                reg = cancellationToken.Register(state => ((StreamPipeWriter)state).Cancel(), this);
+            }
+
+            // Update any buffered data
+            _tail.End += _tailBytesBuffered;
```

Comment:
Do you need to reset `_tailBytesBuffered` in this method? Currently Allocate does it but might be cleaner to also do it here.